RecoEgamma/EgammaElectronAlgos/src/GsfElectronAlgo.cc

```diff
@@ -501,8 +540,18 @@ void GsfElectronAlgo::completeElectrons(reco::GsfElectronCollection& electrons,
                                         edm::Event const& event,
                                         edm::EventSetup const& eventSetup,
                                         const GsfElectronAlgo::HeavyObjectCache* hoc) {
+  auto const& magneticField = eventSetup.getData(magneticFieldToken_);
+  auto const& caloGeometry = eventSetup.getData(caloGeometryToken_);
+  auto const& caloTopology = eventSetup.getData(caloTopologyToken_);
+  auto const& trackerGeometry = eventSetup.getData(trackerGeometryToken_);
+  auto const& ecalSeveretyLevelAlgo = eventSetup.getData(ecalSeveretyLevelAlgoToken_);
+
   checkSetup(eventSetup);
-  auto eventData = beginEvent(event);
+  auto eventData = beginEvent(event, caloGeometry, ecalSeveretyLevelAlgo);
+  double magneticFieldInTesla = magneticField.inTesla(GlobalPoint(0., 0., 0.)).z();
+
+  MultiTrajectoryStateTransform mtsTransform(&trackerGeometry, &magneticField);
+  GsfConstraintAtVertex constraintAtVtx(eventSetup);
```

Comment:
GsfConstraintAtVertex construction seems nontrivial; it was previously done once per IOV.
Please provide profiler report to justify that the cost change is negligible.

Reply:
Nontrivial means here that it gets some things from the EventSetup. My assumption here was that the EventSetup is pretty smart itself, and getting something again that was already obtained in the beginning of the IOV should be basically instantaneous since it returns a cached value (that's also how I realized the manual caching in egamma code is superfluous). See here:

https://twiki.cern.ch/twiki/bin/view/CMSPublic/SWGuideHowToGetDataFromES

"Behind the scenes, the EventSetup system only runs the algorithm used to create the data in a Record the first time that data is requested after a Record has changed its IOV. All subsequent requests will get the cached value."

I can do some profiling anyway, it never hurts of course :)

Reply:
non-trivial means that there is more than just getting the event setup here, at least at two-deep level I found the following (pseudo-code)
``` GsfConstraintAtVertex::GsfConstraintAtVertex () {
 gsfPropagator_ = new GsfPropagatorAdapter(AnalyticalPropagator(magField_, anyDirection)); //copies the temp prop argument
  tipExtrapolator_ = new TransverseImpactPointExtrapolator(*gsfPropagator_); //copies the prop argument
}
```
... after poking deeper now, it seems like the above reduces to some simple initializations and computations of perhaps \~100 operations (pointer assignments, a bunch of "new" for 5-10 words). So, this is probably fine.

Reply:
Thanks a lot for writing this down, I came to the same conclusion back when I made the changes for this PR.